File: src/main/java/org/dukcode/ps/codetree/trail02/chapter06/lesson01/ChallengeTwoNonAdjacentNumbers.java
package org.dukcode.ps.codetree.trail02.chapter06.lesson01;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

/**
 * @see <a
 * href="https://www.codetree.ai/ko/trails/complete/curated-cards/challenge-two-non-adjacent-numbers/description">LINK</a>
 */
public class ChallengeTwoNonAdjacentNumbers {

  private static BufferedReader br;
  private static BufferedWriter bw;
  private static StringTokenizer st;

  public static void main(String[] args) throws IOException {
    br = new BufferedReader(new InputStreamReader(System.in));
    bw = new BufferedWriter(new OutputStreamWriter(System.out));

    int n = Integer.parseInt(br.readLine());
    int[] arr = new int[n];
    st = new StringTokenizer(br.readLine());
    for (int i = 0; i < n; i++) {
      arr[i] = Integer.parseInt(st.nextToken());
    }

    int maxSum = 0;
    for (int i = 0; i < n; i++) {
      for (int j = i + 2; j < n; j++) {
        maxSum = Math.max(maxSum, arr[j] + arr[i]);
      }
    }

    bw.write(String.valueOf(maxSum));

    br.close();
    bw.close();

  }
}

File: src/main/kotlin/org/dukcode/ps/codetree/trail02/chapter06/lesson01/ChallengeTwoNonAdjacentNumbers.kt
package org.dukcode.ps.codetree.trail02.chapter06.lesson01.challengeTwoNonAdjacentNumbers

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.StringTokenizer

private val br = BufferedReader(InputStreamReader(System.`in`))
private val bw = BufferedWriter(OutputStreamWriter(System.out))

/**
 * @see <a href="https://www.codetree.ai/ko/trails/complete/curated-cards/challenge-two-non-adjacent-numbers/description">LINK</a>
 */
fun main() {
    val n = br.readLine().toInt()
    val arr = parseInts()

    val maxPairSum = (0 until n - 2).maxOf { i -> (i + 2 until n).maxOf { j -> arr[i] + arr[j] } }

    bw.write(maxPairSum.toString())

    bw.flush()
}

private fun parseInts() =
    StringTokenizer(br.readLine()).run {
        IntArray(countTokens()) {
            nextToken().toInt()
        }
    }
